service/controller/resource/monitoring/pvcresizingresource/create.go

```diff
@@ -0,0 +1,150 @@
+				// Since Resizing to lower storage is forbidden
+				// Therefore, we replace the PVC and STS
+				// But this will cause data loss
+				fmt.Println("replacePVC..........", pvc.GetName())
+				err = r.replacePVC(ctx, pvc)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+	}
+	r.logger.Debugf(ctx, "ensured pvc resizing")
+
+	return nil
+}
+
+func pvcSelector(clusterID string) labels.Selector {
+	return labels.SelectorFromSet(labels.Set(map[string]string{
+		"app.kubernetes.io/instance":   clusterID,
+		"app.kubernetes.io/managed-by": "prometheus-operator",
+		"app.kubernetes.io/name":       "prometheus",
+		"operator.prometheus.io/name":  clusterID,
+		"prometheus":                   clusterID,
+	}))
+}
+
+func (r *Resource) listPVC(ctx context.Context, clusterID, namespace string) ([]corev1.PersistentVolumeClaim, error) {
+	options := metav1.ListOptions{
+		LabelSelector: pvcSelector(clusterID).String(),
+	}
+	list, err := r.k8sClient.K8sClient().CoreV1().PersistentVolumeClaims(namespace).List(ctx, options)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	if len(list.Items) == 0 {
+		return nil, fmt.Errorf("could not find PVCs")
+	}
+	return list.Items, err
+}
+
+func (r *Resource) resize(ctx context.Context, desiredVolumeSize string, pvc corev1.PersistentVolumeClaim) error {
+
+	namespace := pvc.GetNamespace()
+	clusterID := pvc.GetLabels()["prometheus"]
+
+	newVolumeSize := resource.MustParse(desiredVolumeSize)
+
+	// Patch PVC with the new size
+	pvc.Spec.Resources.Requests["Storage"] = newVolumeSize
+	patch := []byte(fmt.Sprintf(`{"spec": { "resources": { "requests": { "storage": "%v" } } } }`, newVolumeSize.String()))
+	_, err := r.k8sClient.K8sClient().CoreV1().PersistentVolumeClaims(namespace).
+		Patch(ctx, pvc.GetName(), types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Println("PVC patched.....")
+
+	// Delete the sts without the PVC (using orphan)
+	orphan := metav1.DeletePropagationOrphan
+	err = r.k8sClient.K8sClient().AppsV1().StatefulSets(namespace).
+		Delete(ctx, fmt.Sprintf("prometheus-%v", clusterID), metav1.DeleteOptions{PropagationPolicy: &orphan})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Println("Sts deleted.....")
+
+	return nil
+}
+
+func (r *Resource) replacePVC(ctx context.Context, pvc corev1.PersistentVolumeClaim) error {
+
+	namespace := pvc.GetNamespace()
+	clusterID := pvc.GetLabels()["prometheus"]
+
+	patch := []byte(`{"metadata":{"finalizers":null}}`)
+	_, err := r.k8sClient.K8sClient().CoreV1().PersistentVolumeClaims(namespace).
+		Patch(ctx, pvc.GetName(), types.StrategicMergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	err = r.k8sClient.K8sClient().CoreV1().PersistentVolumeClaims(namespace).
+		Delete(ctx, pvc.GetName(), metav1.DeleteOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Println("PVC deleted.....")
```

Comment:
You might have to delete the PV?

Reply:
PVC has StorageClass gp2 defined, so the  PV provisioning and deleting is dynamic